designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,203 @@
+
+### Non-Goals
+
+* Addition of new arbitrary subcommands other than the subcommands that we already support i.e init, create api, and create webhook.
+
+* Discovering plugin binaries that are not locally present on the machine (i.e. exists in a remote repository).
+
+* Discovering flags that each individual plugin would support
+  * Flags would have to be provided by the user as a part of project configuration.
+
+* Providing other options (other than stdin stdout) for inter-process communication between Kubebuilder and the plugins.
+
+### Examples
+
+* `kubebuilder create api --plugins=myexternalplugin/v1`
+  * should scaffold files using the source code of the external plugin as defined in its implementation of the create api method.
+
+* `kubebuilder create api --plugins=myexternalplugin/v1,myexternalplugin/v2`
+  * should scaffold files using the source code of the external plugin as defined in their implementation of the create api method(by respecting the plugin chaining order, i.e. in the order of create api of v1 and then create api of v2 as specified in the layout field in the configuration).
+
+* `kubebuilder create api --plugins=myexternalplugin/v1 --help`
+  * should provide help implementation of the plugin which is not shipped in the binary (source code is present outside of the KB repo).
+
+### How should Kubebuilder discover plugin binaries
+
+Main goal of phase 2 plugins is discovery of out of tree plugins that are external to Kubebuilder and not shipped with the CLI. To be able to run these plugins via the Kubebuilder CLI, it’s necessary for kubebuilder to discover these external plugin binaries and use them as internal ones. I will discuss a few approaches in the order of suitability and preference, with their respective pros and cons.
+
+#### Approach #1
+
+User specified file paths based approach: User will provide a list of file paths for Kubebuilder to discover the plugins in. We will define a variable `KUBEBUILDER_PLUGINS_DIRS` that will take a list of file paths to search for the plugin name. It will also have a default value to search in, in case no file paths are provided. It will search for the plugin name that was provided to the `--plugins` flag in the CLI. Kubebuilder will recursively search for all file paths until the plugin name is found and returns the successful match, and if it doesn’t exist, it returns an error message that the plugin is not found in the provided file paths.
+
+* Pros
+  * This provides flexibility for the user to specify the file paths that the plugin would be placed in and Kubebuilder could discover the binaries in those user specified file paths.
+
+  * No constraints on plugin binary naming or directory placements from the Kubebuilder side.
+
+  * Provides a default value for the plugin directory in case user wants to use that to drop their plugins.
```

Comment:
Could this also be like, when the user provides the `PATH` with a separator, we would look at its location, if not look for the binary in systems's `$PATH` and the last resort would be to figure out the fully qualified path from pwd. Helm kustomize plugin does something like that: https://helm.sh/docs/topics/advanced/#post-rendering

Reply:
I've made a note of this in the EP, in case we want to switch to this approach in the future.